### PR TITLE
Fonglh/duplication framework

### PR DESCRIPTION
**Pull request by fonglh**
I'm done!! If ok I'll squash some of the commits and push again.

### CODE REVIEW

File: spec/libraries/duplicator_spec.rb [hunk 1]
```diff
@@ -0,0 +1,756 @@
...
+      @car4.children << @car3
+      @car5.children << @car4
+    end
+
+    def create_ar_graph
+      #
+      #    --> c22 ---> c23
+      #    |
+      # c21 --> c24 ---> c25
+      #    |              ↑
```
Comment (fonglh) [inline spec/libraries/duplicator_spec.rb]: So can I use non ASCII arrows in the comments?

Comment (lowjoel) [inline spec/libraries/duplicator_spec.rb]: Yah probably better.

File: spec/libraries/duplicator_spec.rb [hunk 2]
```diff
@@ -0,0 +1,756 @@
...
+
+              dup_c22 = dup_c21.children[0]
+              dup_c23 = dup_c22.children[0]
+
+              expect(dup_c21.data).to be(21)
+              expect(dup_c22.data).to be(22)
+              expect(dup_c23.data).to be(23)
+              expect(dup_c21).to_not be(@car21)
+              expect(dup_c22).to_not be(@car22)
+              expect(dup_c23).to_not be (@car23)
```
Comment (lowjoel) [inline spec/libraries/duplicator_spec.rb]: Depending on whether you want identity or equality use eq or be

File: spec/libraries/duplicator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+    temporary_table(:complex_active_records) do |t|
+      t.integer :data
+    end
+
+    temporary_table(:children_parents) do |t|
+      t.integer :children_id, foreign_key: { references: :complex_active_records, primary_key: :id }
+      t.integer :parent_id, foreign_key: { references: :complex_active_records, primary_key: :id }
+    end
+
+    def create_ar_cyclic_graph
```
Comment (fonglh) [inline spec/libraries/duplicator_spec.rb]: Rubocop says ok

File: lib/autoload/duplicator.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  # @param [Enumerable] excluded_objects An Enumerable of objects to be excluded from duplication
+  def initialize(excluded_objects = [])
+    @duplicated_objects = {}  # hash to check what has been duplicated
+    @exclusion_set = excluded_objects.to_set  # set to check what should be excluded
+  end
+
+  # Deep copy +stuff+ and its children. +stuff+ can be a single object or an Enumerable of
+  # objects which must provide a duplicate method which duplicates its children.
+  #
+  # @param[#initialize_duplicate|Enumerable#initialize_duplicate] stuff Either a single object
```
Comment (lowjoel) [inline lib/autoload/duplicator.rb]: I don't think this annotation is right http://yardoc.org/types.html

File: lib/autoload/duplicator.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    stuff = [*stuff] unless stuff.is_a?(Enumerable)
+    stuff.each do |obj|
+      duplicated_stuff << duplicate_object(obj)
+    end
+    duplicated_stuff.length == 1 ? duplicated_stuff[0] : duplicated_stuff
+  end
+
+  private
+
+  # Deep copy +source_object+ and its children. +source_object+ must provide a duplicate
```
Comment (lowjoel) [inline lib/autoload/duplicator.rb]: should monospace `+duplicate+`

File: spec/libraries/duplicator_spec.rb [hunk 6]
```diff
@@ -0,0 +1,760 @@
+require 'rails_helper'
+
+RSpec.describe Duplicator, type: :model do
+  context 'when Plain Old Ruby Objects' do
+    class SimpleObject
+      def initialize(id)
+        @id = id
+      end
+
+      def initialize_duplicate(_duplicator)
```
Comment (lowjoel) [inline spec/libraries/duplicator_spec.rb]: I'd typically just do `_`

File: spec/libraries/duplicator_spec.rb [hunk 7]
```diff
@@ -0,0 +1,760 @@
...
+      #
+      @c22 = ComplexObject.new(22, [])  # assign children later
+      @c23 = ComplexObject.new(23, [@c22])
+      @c21 = ComplexObject.new(21, [@c22, @c23])
+
+      @c22.instance_variable_set(:@children, [@c23])
+    end
+
+    context 'when SimpleObject is duplicated' do
+      before :each do
```
Comment (lowjoel) [inline spec/libraries/duplicator_spec.rb]: `before()`

File: spec/libraries/duplicator_spec.rb [hunk 8]
```diff
@@ -0,0 +1,760 @@
...
+        @id = id
+        @children = children
+      end
+
+      def initialize_duplicate(duplicator)
+        new_children = []
+
+        @children.each do |child|
+          new_child = duplicator.duplicate(child)
+          new_children << new_child unless new_child.nil?
```
Comment (lowjoel) [inline spec/libraries/duplicator_spec.rb]: you can just add all of them to the array, then when you're done call `new_children.compact!`

File: lib/autoload/duplicator.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  # method which duplicates its children.
+  #
+  # @param [#initialize_duplicate] source_object The object to be duplicated.
+  # @return duplicated_object A reference to the duplicated object.
+  def duplicate_object(source_object)
+    if @duplicated_objects.key?(source_object)
+      @duplicated_objects[source_object]
+    elsif !@exclusion_set.include?(source_object)
+      @duplicated_objects[source_object] = source_object.dup
+      source_object.initialize_duplicate(self)
```
Comment (lowjoel) [inline lib/autoload/duplicator.rb]: @fonglh I have a question here, which might break your entire PR. Should the duplicate initialise itself, or the source object initialise the duplicate?

File: spec/libraries/duplicator_spec.rb [hunk 10]
```diff
@@ -0,0 +1,760 @@
...
+        end
+
+        # example if object has no accessor for the children
+        # self is already duplicated by Duplicator. Find the duplicated version
+        duplicator.duplicate(self).instance_variable_set(:@children, new_children)
```
Comment (lowjoel) [inline spec/libraries/duplicator_spec.rb]: If self was the duplicated object, then this would be trivial.

File: lib/autoload/duplicator.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+  # Deep copy +source_object+ and its children. +source_object+ must provide a duplicate
+  # method which duplicates its children.
+  #
+  # @param [#initialize_duplicate] source_object The object to be duplicated.
+  # @return duplicated_object A reference to the duplicated object.
+  def duplicate_object(source_object)
+    @duplicated_objects.fetch(source_object) do |key|
+      if !@exclusion_set.include?(key)
+        @duplicated_objects[key] = source_object.dup
```
Comment (lowjoel) [inline lib/autoload/duplicator.rb]:
```ruby
source_object.dup do |duplicate|
  @duplicated_objects[key] = duplicate
  duplicate.initialize_duplicate(self, key)
end
```

File: spec/libraries/duplicator_spec.rb [hunk 12]
```diff
@@ -0,0 +1,763 @@
...
+    # ComplexObject has children
+    class ComplexObject
+      attr_reader :children
+
+      def initialize(id, children)
+        @id = id
+        @children = children
+      end
+
+      def initialize_duplicate(duplicator, other)
```
Comment (fonglh) [inline spec/libraries/duplicator_spec.rb]: This doesn't seem right. `other` is used in the method.

Comment (fonglh) [inline spec/libraries/duplicator_spec.rb]: @lowjoel

Comment (lowjoel) [inline spec/libraries/duplicator_spec.rb]: Ignore it.

**Comment by lowjoel** [comment]
@fonglh I'm generally okay with this PR, please squash it down to a smaller number of commits, each only dealing with a significant addition to the PR.

**Comment by fonglh** [comment]
Squashed.

**Comment by lowjoel** [comment]
Time for the registration bit!

**Pull request merged**